libp2p-xtra/examples/hello_world_dialer.rs

```diff
@@ -37,11 +39,22 @@ async fn main() -> Result<()> {
         .create(None)
         .spawn_global();
 
-    endpoint_addr
-        .send(Connect(opts.multiaddr.clone()))
-        .await
-        .unwrap()
-        .unwrap();
+    let dialer_constructor = {
+        // XXX: Can this be done nicer???
+        let connect_addr = opts.multiaddr.clone();
+        let endpoint_addr = endpoint_addr.clone();
+        move || {
+            let connect_addr = connect_addr.clone();
+            let endpoint_addr = endpoint_addr.clone();
+            dialer::Actor::new(endpoint_addr, connect_addr)
+        }
+    };
+
+    let (supervisor, _dialer_actor) = supervisor::Actor::with_policy(
+        dialer_constructor,
+        |_: &dialer::Error| true, // always restart dialer actor
```

Comment:
yes, this is the only error that is emitted! do you think it would be less confusing if I just create a supervisor actor without any explicit policy? (I believe it will then restart on *all* the errors?)

Reply:
Huh, I thought the type hint was superfluous. The `dialer::Actor` can _only_ emit a `dialer::Error`, so that's the only thing the `supervisor::Actor` can expect here. Therefore, this actor will always be restarted. Any flaws in my logic?

Reply:
unfortunately, it seems that if we want actors to be flexible, we cannot use the simple constructor anymore :( might look into the supervisor later on.

Reply:
> unfortunately, it seems that if we want actors to be flexible, we cannot use the simple constructor anymore :( might look into the supervisor later on.

What do you mean by flexible?

> Huh, I thought the type hint was superfluous. The dialer::Actor can only emit a dialer::Error, so that's the only thing the supervisor::Actor can expect here. Therefore, this actor will always be restarted. Any flaws in my logic?

I was wrong in the sense that the type hint is needed, but it's true that the error emitted by the supervised actor when stopping is predictable (and always the same).

Reply:
flexible - meaning that one can use a supervisor with a custom policy (turns out if we define a custom error, we can't use `new()`, we have to rely on `with_policy()` - and the decision should be relying on the users of the API, not baked inside particular actors.

yup, type hint is needed indeed 😮‍💨 Errors are different, depending on the way the actor stops. the consumers of this API could easily say "don't try to restart if you've been disconnected" or perhaps "wait for X seconds before you attempt to re-dial when you see such-or-such error. 
In this example, it just restarts straight away.